game-core/src/main/java/games/strategy/triplea/ai/tree/BattleStep.java

```diff
@@ -0,0 +1,1216 @@
+        this.badProbability =
+            1.0 - (this.winProbability + this.loseProbability + this.tieProbability);
+        // overwrite the cached one with this that has a result
+        nodeCache.put(units, this);
+        // System.out.println("Recurse: " + this);
+      }
+      this.hasResult = true;
+      return true;
+    }
+
+    if (round >= MAX_ROUNDS) {
+      // some branches go for ever
+      // an example is where no one hits each other
+      // treat any of these branches as a tie
+      System.out.println("Reached max rounds");
+      hasResult = true;
+      badProbability = 1.0;
+      // System.out.println("Max: " + this);
+      return true;
+    }
+
+    nodeCache.put(units, this);
+    return false;
+  }
+
+  private void calculateSiblingResultsForRecursiveEnd(
+      final List<BattleStep> steps, final double[] siblingResults, final StepUnits siblingUnits) {
+    for (final BattleStep level : steps) {
+      if (level.hasResult) {
+        continue;
+      }
+      siblingResults[4] *= level.probability;
+      siblingResults[0] += siblingResults[4] * level.winProbability;
+      siblingResults[1] += siblingResults[4] * level.loseProbability;
+      siblingResults[2] += siblingResults[4] * level.tieProbability;
+      siblingResults[3] += siblingResults[4] * level.averageRounds;
+      if (level.units.getType() != Type.AA_ATTACKER) {
+        if (siblingUnits.getPlayer().equals(level.averageUnits.getPlayer())) {
+          siblingUnits.updateUnitChances(level.averageUnits.swapSides(), siblingResults[4]);
+        } else {
+          siblingUnits.updateUnitChances(level.averageUnits, siblingResults[4]);
+        }
+        calculateSiblingResultsForRecursiveEnd(level.children, siblingResults, siblingUnits);
+      }
+    }
+  }
+
+  /**
+   * Figures out all the probabilities of rolling dice with a specific set of powers and amounts.
+   *
+   * @param rollData The details about the dice available to roll
+   * @return List of possible probabilities from rolling 0 dice to rolling rollData.totalRolls})
+   */
+  List<Double> calculateHitProbabilities(final RollData rollData) {
+    if (rollData.totalRolls == 0) {
+      return List.of();
+    }
+    final List<Integer> dicePowersList = new ArrayList<>(rollData.rollsByDicePower.keySet());
+    Collections.sort(dicePowersList);
+    final int maxPower = dicePowersList.get(dicePowersList.size() - 1);
+    final int[] dicePowers = new int[dicePowersList.size()];
+
+    final Map<Integer, double[]> calculatedProbabilityCache =
+        this.calculatedProbabilityCache.computeIfAbsent(
+            rollData.maxDiceSides,
+            key -> new HashMap<>(rollData.maxDiceSides * rollData.totalRolls));
+
+    final int[] diceAvailableGroupedByPower = new int[maxPower + 1];
+    for (int dicePowerIndex = 0; dicePowerIndex < dicePowersList.size(); dicePowerIndex++) {
+      final int dicePower = dicePowersList.get(dicePowerIndex);
+
+      dicePowers[dicePowerIndex] = dicePower;
```

Comment:
You can use `final int[] dicePowers = dicePowersList.toArray(new int[0]);` to convert a List to an array.
Because it's an arraylist implementation copying the whole array might be slightly faster than iterating

Reply:
I already have to loop through the array to calculate `diceAvailableGroupedByPower` so doing two things in the same loop vs two different loops is better in this situation.  This function is called a lot.